File: engine/src/analyze/codeGraph.ts
import {
  type CodeGraph as BaseCodeGraph,
  type CodeGraphEdge,
  type CodeGraphEdgeKind,
  type CodeGraphNode,
  fileNode,
  type FileSymbols,
  type Symbol,
} from "../models.js";
import { ImportResolver } from "./importResolver.js";

export interface CodeGraph extends BaseCodeGraph {
  externalDependencies: Record<string, string[]>;
}

export function symbolNode(symbol: Symbol, path: string): CodeGraphNode {
  return {
    id: `symbol:${path}#${symbol.name}`,
    kind: "symbol",
    filePath: path,
    symbolName: symbol.name,
  };
}

function edgeKey(edge: CodeGraphEdge): string {
  return `${edge.from.id}\u0000${edge.to.id}\u0000${edge.kind}`;
}

function matchesKinds(kind: CodeGraphEdgeKind, kinds: Set<CodeGraphEdgeKind> | undefined): boolean {
  return kinds === undefined || kinds.has(kind);
}

export function outboundEdges(
  graph: CodeGraph,
  node: CodeGraphNode,
  kinds?: Set<CodeGraphEdgeKind>,
): CodeGraphEdge[] {
  return graph.edges.filter(
    (edge) => edge.from.id === node.id && matchesKinds(edge.kind, kinds),
  );
}

export function inboundEdges(
  graph: CodeGraph,
  node: CodeGraphNode,
  kinds?: Set<CodeGraphEdgeKind>,
): CodeGraphEdge[] {
  return graph.edges.filter((edge) => edge.to.id === node.id && matchesKinds(edge.kind, kinds));
}

export function neighbors(
  graph: CodeGraph,
  node: CodeGraphNode,
  kinds?: Set<CodeGraphEdgeKind>,
): CodeGraphNode[] {
  const seen = new Set<string>();
  const out: CodeGraphNode[] = [];
  for (const edge of outboundEdges(graph, node, kinds)) {
    if (!seen.has(edge.to.id)) {
      seen.add(edge.to.id);
      out.push(edge.to);
    }
  }
  return out;
}

export function transitiveClosure(
  graph: CodeGraph,
  node: CodeGraphNode,
  maxDepth: number,
  kinds?: Set<CodeGraphEdgeKind>,
): CodeGraphNode[] {
  if (maxDepth <= 0) {
    return [];
  }

  const visited = new Map<string, CodeGraphNode>();
  let frontier: CodeGraphNode[] = [node];

  for (let depth = 0; depth < maxDepth; depth += 1) {
    const nextById = new Map<string, CodeGraphNode>();
    for (const current of frontier) {
      for (const neighbor of neighbors(graph, current, kinds)) {
        if (neighbor.id !== node.id && !visited.has(neighbor.id)) {
          nextById.set(neighbor.id, neighbor);
        }
      }
    }

    if (nextById.size === 0) {
      break;
    }

    for (const [id, value] of nextById) {
      visited.set(id, value);
    }
    frontier = [...nextById.values()];
  }

  return [...visited.values()];
}

function escapeRegExp(value: string): string {
  return value.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}

function regexMatch(pattern: string, contents: string): boolean {
  try {
    return new RegExp(pattern).test(contents);
  } catch {
    return false;
  }
}

function containsCall(symbolName: string, contents: string): boolean {
  return regexMatch(String.raw`\b${escapeRegExp(symbolName)}\s*\(`, contents);
}

function containsReference(symbolName: string, contents: string): boolean {
  return regexMatch(String.raw`\b${escapeRegExp(symbolName)}\b`, contents);
}

function containsInheritance(symbolName: string, contents: string): boolean {
  const escaped = escapeRegExp(symbolName);
  return (
    regexMatch(String.raw`\bextends\s+${escaped}\b`, contents) ||
    regexMatch(String.raw`:\s*${escaped}\b`, contents)
  );
}

function containsImplementation(symbolName: string, contents: string): boolean {
  return regexMatch(String.raw`\bimplements\s+${escapeRegExp(symbolName)}\b`, contents);
}

interface ExportedSymbol {
  filePath: string;
  symbol: Symbol;
}

function exportedSymbolIndex(files: FileSymbols[]): ExportedSymbol[] {
  const out: ExportedSymbol[] = [];
  for (const file of files) {
    const exportedNames = new Set(file.exports.map((ref) => ref.name));
    for (const symbol of file.symbols) {
      if (exportedNames.has(symbol.name) || file.exports.length === 0) {
        out.push({ filePath: file.path, symbol });
      }
    }
  }
  return out;
}

export class CodeGraphBuilder {
  private readonly resolver: ImportResolver;

  constructor(resolver: ImportResolver) {
    this.resolver = resolver;
  }

  build(files: FileSymbols[], contentsByPath: Record<string, string> = {}): CodeGraph {
    const nodesById = new Map<string, CodeGraphNode>();
    const edgesByKey = new Map<string, CodeGraphEdge>();
    const externalDependencies = new Map<string, Set<string>>();

    const insertNode = (node: CodeGraphNode): void => {
      if (!nodesById.has(node.id)) {
        nodesById.set(node.id, node);
      }
    };
    const insertEdge = (edge: CodeGraphEdge): void => {
      const key = edgeKey(edge);
      if (!edgesByKey.has(key)) {
        edgesByKey.set(key, edge);
      }
    };

    const symbolNodesByPath = new Map<string, CodeGraphNode[]>();
    for (const file of files) {
      symbolNodesByPath.set(
        file.path,
        file.symbols.map((symbol) => symbolNode(symbol, file.path)),
      );
    }
    const exportedSymbols = exportedSymbolIndex(files);

    for (const file of files) {
      const node = fileNode(file.path);
      insertNode(node);
      for (const symbolNodeValue of symbolNodesByPath.get(file.path) ?? []) {
        insertNode(symbolNodeValue);
      }

      for (const importRef of file.imports) {
        const targetPath =
          this.resolver.resolve(importRef.moduleSpecifier, file.path) ?? importRef.resolvedTarget;
        if (!targetPath) {
          if (!externalDependencies.has(file.path)) {
            externalDependencies.set(file.path, new Set());
          }
          externalDependencies.get(file.path)?.add(importRef.moduleSpecifier);
          continue;
        }

        const targetNode = fileNode(targetPath);
        insertNode(targetNode);
        insertEdge({ from: node, to: targetNode, kind: "import" });
      }

      const contents = contentsByPath[file.path];
      if (contents === undefined) {
        continue;
      }

      for (const exported of exportedSymbols) {
        if (exported.filePath === file.path) {
          continue;
        }

        const exportedNode = symbolNode(exported.symbol, exported.filePath);
        insertNode(exportedNode);

        if (containsCall(exported.symbol.name, contents)) {
          insertEdge({ from: node, to: exportedNode, kind: "call" });
        } else if (containsReference(exported.symbol.name, contents)) {
          insertEdge({ from: node, to: exportedNode, kind: "reference" });
        }

        if (containsInheritance(exported.symbol.name, contents)) {
          insertEdge({ from: node, to: exportedNode, kind: "inherit" });
        }

        if (containsImplementation(exported.symbol.name, contents)) {
          insertEdge({ from: node, to: exportedNode, kind: "implements" });
        }
      }
    }

    const externalDependenciesRecord: Record<string, string[]> = {};
    for (const [path, deps] of externalDependencies) {
      externalDependenciesRecord[path] = [...deps];
    }

    return {
      nodes: [...nodesById.values()],
      edges: [...edgesByKey.values()],
      externalDependencies: externalDependenciesRecord,
    };
  }
}

File: app/Sources/GunkApp/Analyze/CodeGraph.swift
import Foundation

struct CodeGraphNode: Equatable, Hashable, Sendable {
  enum Kind: Equatable, Hashable, Sendable {
    case file
    case symbol(Symbol.Kind)
  }

  let id: String
  let kind: Kind
  let filePath: String
  let symbolName: String?

  static func file(_ path: String) -> CodeGraphNode {
    CodeGraphNode(id: "file:\(path)", kind: .file, filePath: path, symbolName: nil)
  }

  static func symbol(_ symbol: Symbol, in path: String) -> CodeGraphNode {
    CodeGraphNode(
      id: "symbol:\(path)#\(symbol.name)",
      kind: .symbol(symbol.kind),
      filePath: path,
      symbolName: symbol.name
    )
  }
}

enum CodeGraphEdgeKind: String, Equatable, Hashable, Sendable {
  case call
  case `import`
  case implements
  case inherit
  case reference
}

struct CodeGraphEdge: Equatable, Hashable, Sendable {
  let from: CodeGraphNode
  let to: CodeGraphNode
  let kind: CodeGraphEdgeKind
}

struct CodeGraph: Equatable, Sendable {
  let nodes: Set<CodeGraphNode>
  let edges: Set<CodeGraphEdge>
  let externalDependencies: [String: Set<String>]

  func neighbors(of node: CodeGraphNode, kinds: Set<CodeGraphEdgeKind>? = nil) -> Set<CodeGraphNode> {
    Set(outboundEdges(from: node, kinds: kinds).map(\.to))
  }

  func outboundEdges(from node: CodeGraphNode, kinds: Set<CodeGraphEdgeKind>? = nil) -> Set<CodeGraphEdge> {
    Set(edges.filter { edge in
      edge.from == node && (kinds == nil || kinds?.contains(edge.kind) == true)
    })
  }

  func inboundEdges(to node: CodeGraphNode, kinds: Set<CodeGraphEdgeKind>? = nil) -> Set<CodeGraphEdge> {
    Set(edges.filter { edge in
      edge.to == node && (kinds == nil || kinds?.contains(edge.kind) == true)
    })
  }

  func transitiveClosure(from node: CodeGraphNode, maxDepth: Int, kinds: Set<CodeGraphEdgeKind>? = nil) -> Set<CodeGraphNode> {
    guard maxDepth > 0 else {
      return []
    }

    var visited: Set<CodeGraphNode> = []
    var frontier: Set<CodeGraphNode> = [node]

    for _ in 0..<maxDepth {
      let next = Set(frontier.flatMap { neighbors(of: $0, kinds: kinds) }).subtracting(visited).subtracting([node])
      guard !next.isEmpty else {
        break
      }

      visited.formUnion(next)
      frontier = next
    }

    return visited
  }
}

struct CodeGraphBuilder: Sendable {
  private let resolver: ImportResolver

  init(resolver: ImportResolver) {
    self.resolver = resolver
  }

  func build(files: [FileSymbols], contentsByPath: [String: String] = [:]) -> CodeGraph {
    var nodes: Set<CodeGraphNode> = []
    var edges: Set<CodeGraphEdge> = []
    var externalDependencies: [String: Set<String>] = [:]

    let symbolNodesByPath = Dictionary(
      uniqueKeysWithValues: files.map { file in
        (file.path, file.symbols.map { CodeGraphNode.symbol($0, in: file.path) })
      }
    )
    let exportedSymbols = exportedSymbolIndex(files: files)

    for file in files {
      let fileNode = CodeGraphNode.file(file.path)
      nodes.insert(fileNode)
      nodes.formUnion(symbolNodesByPath[file.path] ?? [])

      for importRef in file.imports {
        guard let targetPath = resolver.resolve(importRef.moduleSpecifier, from: file.path) ?? importRef.resolvedTarget else {
          externalDependencies[file.path, default: []].insert(importRef.moduleSpecifier)
          continue
        }

        let targetNode = CodeGraphNode.file(targetPath)
        nodes.insert(targetNode)
        edges.insert(CodeGraphEdge(from: fileNode, to: targetNode, kind: .import))
      }

      guard let contents = contentsByPath[file.path] else {
        continue
      }

      for exported in exportedSymbols where exported.filePath != file.path {
        let symbolNode = CodeGraphNode.symbol(exported.symbol, in: exported.filePath)
        nodes.insert(symbolNode)

        if containsCall(to: exported.symbol.name, in: contents) {
          edges.insert(CodeGraphEdge(from: fileNode, to: symbolNode, kind: .call))
        } else if containsReference(to: exported.symbol.name, in: contents) {
          edges.insert(CodeGraphEdge(from: fileNode, to: symbolNode, kind: .reference))
        }

        if containsInheritance(of: exported.symbol.name, in: contents) {
          edges.insert(CodeGraphEdge(from: fileNode, to: symbolNode, kind: .inherit))
        }

        if containsImplementation(of: exported.symbol.name, in: contents) {
          edges.insert(CodeGraphEdge(from: fileNode, to: symbolNode, kind: .implements))
        }
      }
    }

    return CodeGraph(nodes: nodes, edges: edges, externalDependencies: externalDependencies)
  }

  private func exportedSymbolIndex(files: [FileSymbols]) -> [(filePath: String, symbol: Symbol)] {
    files.flatMap { file in
      let exportedNames = Set(file.exports.map(\.name))
      return file.symbols
        .filter { exportedNames.contains($0.name) || file.exports.isEmpty }
        .map { (file.path, $0) }
    }
  }

  private func containsCall(to symbolName: String, in contents: String) -> Bool {
    regexMatch(#"\b\#(NSRegularExpression.escapedPattern(for: symbolName))\s*\("#, in: contents)
  }

  private func containsReference(to symbolName: String, in contents: String) -> Bool {
    regexMatch(#"\b\#(NSRegularExpression.escapedPattern(for: symbolName))\b"#, in: contents)
  }

  private func containsInheritance(of symbolName: String, in contents: String) -> Bool {
    let escaped = NSRegularExpression.escapedPattern(for: symbolName)
    return regexMatch(#"\bextends\s+\#(escaped)\b"#, in: contents)
      || regexMatch(#":\s*\#(escaped)\b"#, in: contents)
  }

  private func containsImplementation(of symbolName: String, in contents: String) -> Bool {
    regexMatch(#"\bimplements\s+\#(NSRegularExpression.escapedPattern(for: symbolName))\b"#, in: contents)
  }

  private func regexMatch(_ pattern: String, in contents: String) -> Bool {
    guard let regex = try? NSRegularExpression(pattern: pattern) else {
      return false
    }

    let range = NSRange(location: 0, length: (contents as NSString).length)
    return regex.firstMatch(in: contents, range: range) != nil
  }
}
